sdks/python/apache_beam/typehints/decorators.py

```diff
@@ -228,18 +234,30 @@ class IOTypeHints(object):
       May be None. The list and dict correspond to args and kwargs.
     output_types: (tuple, dict) List of typing types, and an optional dictionary
       (unused). Only the first element of the list is used. May be None.
+    origin: (List[str]) Stack of tracebacks of method calls used to create this
+      instance.
   """
-  __slots__ = ('input_types', 'output_types')
 
-  def __init__(self,
-               input_types=None,  # type: Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]
-               output_types=None  # type: Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]
-              ):
-    self.input_types = input_types
-    self.output_types = output_types
+  traceback_limit = 5
 
-  @staticmethod
-  def from_callable(fn):
+  @classmethod
+  def _make_traceback(cls, base):  # type: (Optional[IOTypeHints]) -> List[str]
+    # Omit this method and the IOTypeHints method that called it.
+    num_frames_skip = 2
+    tb = traceback.format_stack(limit=cls.traceback_limit + num_frames_skip)
+    tb_lines = 'TH>' + ''.join(tb[:-num_frames_skip]).replace('\n', '\nTH>')
+
+    res = [tb_lines + '\nbased on: ' + str(base)]
+    if base is not None:
+      res += base.origin
+    return res
+
+  @classmethod
+  def empty(cls):
+    return IOTypeHints(None, None, [])
```

Comment:
would be nice to add a type annotation to this:

```python
  def empty(cls):
    # type: () -> IOTypeHints
    return IOTypeHints(None, None, [])
```

Reply:
done